ext/tag_list/main.php

```diff
@@ -404,25 +404,21 @@ private function add_related_block(Page $page, Image $image)
 
         $query = "
 			SELECT t3.tag AS tag, t3.count AS calc_count, it3.tag_id
-			FROM
-				image_tags AS it1,
-				image_tags AS it2,
-				image_tags AS it3,
-				tags AS t1,
-				tags AS t3
+            FROM image_tags AS it1 -- Starting image's tags                
+                INNER JOIN tags AS t1 ON t1.id = it1.tag_id AND t1.tag NOT LIKE 'tagme%'
+                -- Get images with the same tags as the starting image
+                INNER JOIN image_tags AS it2 ON it1.tag_id=it2.tag_id
+                -- Get the tags from those other images except the same as the starting tags 
+                INNER JOIN image_tags AS it3 ON it2.image_id=it3.image_id 
+                LEFT JOIN image_tags it4 ON it4.image_id = it1.image_id AND it4.tag_id = it3.tag_id
+                INNER JOIN tags AS t3 ON t3.id = it3.tag_id AND t3.tag NOT LIKE 'tagme%' 
```

Comment:
This runs slower than the original for me o_O 250 vs 500ms on a random example image (adding tagme-wildcard support into the original only adds 10ms, so that doesn't seem like the main thing)

Original:
```
=> explain analyse SELECT t3.tag AS tag, t3.count AS calc_count, it3.tag_id
            FROM
                image_tags AS it1,
                image_tags AS it2,
                image_tags AS it3,
                tags AS t1,
                tags AS t3
            WHERE
                it1.image_id=245368 AND it1.tag_id=it2.tag_id
                AND it2.image_id=it3.image_id
                AND t1.tag != 'tagme'
                AND t3.tag != 'tagme'
                AND t1.id = it1.tag_id
                AND t3.id = it3.tag_id
            GROUP BY it3.tag_id, t3.tag, t3.count
            ORDER BY calc_count DESC
            LIMIT 10;
                                                                                      QUERY PLAN
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 Limit  (cost=112322.65..112322.68 rows=10 width=20) (actual time=246.339..246.345 rows=10 loops=1)
   ->  Sort  (cost=112322.65..114323.73 rows=800433 width=20) (actual time=246.338..246.338 rows=10 loops=1)
         Sort Key: t3.count DESC
         Sort Method: top-N heapsort  Memory: 26kB
         ->  HashAggregate  (cost=87021.25..95025.58 rows=800433 width=20) (actual time=242.116..245.890 rows=3148 loops=1)
               Group Key: t3.count, it3.tag_id, t3.tag
               ->  Hash Join  (cost=7141.79..81018.01 rows=800433 width=20) (actual time=59.116..205.044 rows=132949 loops=1)
                     Hash Cond: (it3.tag_id = t3.id)
                     ->  Nested Loop  (cost=1.73..71776.75 rows=800437 width=4) (actual time=0.046..116.141 rows=133376 loops=1)
                           ->  Nested Loop  (cost=1.29..232.39 rows=106638 width=4) (actual time=0.039..17.900 rows=31082 loops=1)
                                 Join Filter: (it1.tag_id = it2.tag_id)
                                 ->  Nested Loop  (cost=0.85..72.10 rows=8 width=8) (actual time=0.026..0.040 rows=3 loops=1)
                                       ->  Index Only Scan using image_tags_image_id_key on image_tags it1  (cost=0.43..4.57 rows=8 width=4) (actual time=0.014..0.015 rows=3 loops=1)
                                             Index Cond: (image_id = 245368)
                                             Heap Fetches: 3
                                       ->  Index Scan using tags_pkey on tags t1  (cost=0.42..8.44 rows=1 width=4) (actual time=0.006..0.006 rows=1 loops=3)
                                             Index Cond: (id = it1.tag_id)
                                             Filter: (tag <> 'tagme'::text)
                                 ->  Index Scan using image_tags_tag_id_idx on image_tags it2  (cost=0.43..12.60 rows=595 width=8) (actual time=0.009..4.689 rows=10361 loops=3)
                                       Index Cond: (tag_id = t1.id)
                           ->  Index Only Scan using image_tags_image_id_key on image_tags it3  (cost=0.43..0.59 rows=8 width=8) (actual time=0.002..0.003 rows=4 loops=31082)
                                 Index Cond: (image_id = it2.image_id)
                                 Heap Fetches: 65891
                     ->  Hash  (cost=4376.54..4376.54 rows=221082 width=20) (actual time=58.816..58.816 rows=221321 loops=1)
                           Buckets: 262144  Batches: 1  Memory Usage: 13613kB
                           ->  Seq Scan on tags t3  (cost=0.00..4376.54 rows=221082 width=20) (actual time=0.013..25.065 rows=221321 loops=1)
                                 Filter: (tag <> 'tagme'::text)
                                 Rows Removed by Filter: 1
 Planning Time: 1.262 ms
 Execution Time: 249.340 ms
```

New:
```
=> explain analyse SELECT t3.tag AS tag, t3.count AS calc_count, it3.tag_id
            FROM image_tags AS it1 -- Starting image's tags
                INNER JOIN tags AS t1 ON t1.id = it1.tag_id AND t1.tag NOT LIKE 'tagme%'
                -- Get images with the same tags as the starting image
                INNER JOIN image_tags AS it2 ON it1.tag_id=it2.tag_id
                -- Get the tags from those other images except the same as the starting tags
                INNER JOIN image_tags AS it3 ON it2.image_id=it3.image_id
                LEFT JOIN image_tags it4 ON it4.image_id = it1.image_id AND it4.tag_id = it3.tag_id
                INNER JOIN tags AS t3 ON t3.id = it3.tag_id AND t3.tag NOT LIKE 'tagme%'
            WHERE
                it1.image_id=245368 GROUP BY it3.tag_id, t3.tag, t3.count
            ORDER BY calc_count DESC
            LIMIT 10;
                                                                                      QUERY PLAN
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 Limit  (cost=1725.26..1725.28 rows=10 width=20) (actual time=509.242..509.245 rows=10 loops=1)
   ->  Sort  (cost=1725.26..1731.86 rows=2642 width=20) (actual time=509.241..509.241 rows=10 loops=1)
         Sort Key: t3.count DESC
         Sort Method: top-N heapsort  Memory: 26kB
         ->  HashAggregate  (cost=1641.74..1668.16 rows=2642 width=20) (actual time=508.271..508.808 rows=3148 loops=1)
               Group Key: t3.count, it3.tag_id, t3.tag
               ->  Nested Loop  (cost=2.15..1621.93 rows=2642 width=20) (actual time=0.035..459.210 rows=132949 loops=1)
                     ->  Nested Loop  (cost=1.73..456.65 rows=2642 width=4) (actual time=0.030..129.755 rows=133376 loops=1)
                           ->  Nested Loop  (cost=1.29..220.49 rows=352 width=8) (actual time=0.025..19.289 rows=31082 loops=1)
                                 ->  Nested Loop  (cost=0.85..72.10 rows=8 width=12) (actual time=0.016..0.028 rows=3 loops=1)
                                       ->  Index Only Scan using image_tags_image_id_key on image_tags it1  (cost=0.43..4.57 rows=8 width=8) (actual time=0.008..0.010 rows=3 loops=1)
                                             Index Cond: (image_id = 245368)
                                             Heap Fetches: 3
                                       ->  Index Scan using tags_pkey on tags t1  (cost=0.42..8.44 rows=1 width=4) (actual time=0.004..0.004 rows=1 loops=3)
                                             Index Cond: (id = it1.tag_id)
                                             Filter: (tag !~~ 'tagme%'::text)
                                 ->  Index Scan using image_tags_tag_id_idx on image_tags it2  (cost=0.43..12.60 rows=595 width=8) (actual time=0.009..4.963 rows=10361 loops=3)
                                       Index Cond: (tag_id = t1.id)
                           ->  Index Only Scan using image_tags_image_id_key on image_tags it3  (cost=0.43..0.59 rows=8 width=8) (actual time=0.002..0.003 rows=4 loops=31082)
                                 Index Cond: (image_id = it2.image_id)
                                 Heap Fetches: 65891
                     ->  Index Scan using tags_pkey on tags t3  (cost=0.42..0.44 rows=1 width=20) (actual time=0.002..0.002 rows=1 loops=133376)
                           Index Cond: (id = it3.tag_id)
                           Filter: (tag !~~ 'tagme%'::text)
                           Rows Removed by Filter: 0
 Planning Time: 1.272 ms
 Execution Time: 509.295 ms
```

Reply:
Interesting, my analyze takes advantage of image_tags_image_id_tag_id_key while yours is just using  image_tags_image_id_key. Are you missing image_tags_image_id_tag_id_key on your table?

Reply:
I don't have that one, though I do have UNIQUE BTREE key across image_tags(image_id, tag_id):

```
=> \d image_tags;
              Table "public.image_tags"
  Column  │  Type   │ Collation │ Nullable │ Default
──────────┼─────────┼───────────┼──────────┼─────────
 image_id │ integer │           │ not null │
 tag_id   │ integer │           │ not null │
Indexes:
    "image_tags_image_id_key" UNIQUE CONSTRAINT, btree (image_id, tag_id)
    "image_tags_image_id_idx" btree (image_id) CLUSTER
    "image_tags_tag_id_idx" btree (tag_id)
Foreign-key constraints:
    "image_tags_image_id" FOREIGN KEY (image_id) REFERENCES images(id) ON DELETE CASCADE
    "image_tags_tag_id" FOREIGN KEY (tag_id) REFERENCES tags(id) ON DELETE RESTRICT
```

Reply:
I see, your image_tags_image_id_key is the same as my image_tags_image_id_tag_id_key. Must have been a difference in how the indexes were named by postgres or something.

I also see that your analyzes are using image_tags_tag_id_idx rather than the image_tags_tag_id_image_id_idx index that was added in one of these commits. I just dropped image_tags_tag_id_image_id_idx  it from my database and it increased the query time by 400ms, so I'd recommend adding it and re-analyzing to see if it has an impact.

Reply:
Added that index, and it's still slower on my data :S Could you do an explain analyse / ` \d image_tags` on your database to compare?
```
> explain analyse SELECT t3.tag AS tag, t3.count AS calc_count, it3.tag_id
                                  FROM
                                      image_tags AS it1,
                                      image_tags AS it2,
                                      image_tags AS it3,
                                      tags AS t1,
                                      tags AS t3
                                  WHERE
                                      it1.image_id=245368 AND it1.tag_id=it2.tag_id
                                      AND it2.image_id=it3.image_id
                                      AND t1.tag != 'tagme'
                                      AND t3.tag != 'tagme'
                                      AND t1.id = it1.tag_id
                                      AND t3.id = it3.tag_id
                                  GROUP BY it3.tag_id, t3.tag, t3.count
                                  ORDER BY calc_count DESC
                                  LIMIT 10;

+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| QUERY PLAN                                                                                                                                                                                    |
|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Limit  (cost=116519.43..116519.46 rows=10 width=19) (actual time=248.083..248.088 rows=10 loops=1)                                                                                            |
|   ->  Sort  (cost=116519.43..118571.31 rows=820752 width=19) (actual time=248.081..248.082 rows=10 loops=1)                                                                                   |
|         Sort Key: t3.count DESC                                                                                                                                                               |
|         Sort Method: top-N heapsort  Memory: 25kB                                                                                                                                             |
|         ->  HashAggregate  (cost=90575.76..98783.28 rows=820752 width=19) (actual time=244.100..247.658 rows=3223 loops=1)                                                                    |
|               Group Key: t3.count, it3.tag_id, t3.tag                                                                                                                                         |
|               ->  Hash Join  (cost=7253.44..84420.12 rows=820752 width=19) (actual time=59.350..206.896 rows=136565 loops=1)                                                                  |
|                     Hash Cond: (it3.tag_id = t3.id)                                                                                                                                           |
|                     ->  Nested Loop  (cost=1.73..75013.88 rows=820756 width=4) (actual time=0.064..117.785 rows=136996 loops=1)                                                               |
|                           ->  Nested Loop  (cost=1.29..245.40 rows=109345 width=4) (actual time=0.056..17.345 rows=31872 loops=1)                                                             |
|                                 Join Filter: (it1.tag_id = it2.tag_id)                                                                                                                        |
|                                 ->  Nested Loop  (cost=0.85..87.03 rows=8 width=8) (actual time=0.041..0.055 rows=3 loops=1)                                                                  |
|                                       ->  Index Only Scan using image_tags_image_id_key on image_tags it1  (cost=0.43..19.51 rows=8 width=4) (actual time=0.024..0.026 rows=3 loops=1)        |
|                                             Index Cond: (image_id = 245368)                                                                                                                   |
|                                             Heap Fetches: 3                                                                                                                                   |
|                                       ->  Index Scan using tags_pkey on tags t1  (cost=0.42..8.44 rows=1 width=4) (actual time=0.007..0.007 rows=1 loops=3)                                   |
|                                             Index Cond: (id = it1.tag_id)                                                                                                                     |
|                                             Filter: (tag <> 'tagme'::text)                                                                                                                    |
|                                 ->  Index Only Scan using image_tags_tag_id_image_id_idx on image_tags it2  (cost=0.43..12.17 rows=610 width=8) (actual time=0.011..4.507 rows=10624 loops=3) |
|                                       Index Cond: (tag_id = t1.id)                                                                                                                            |
|                                       Heap Fetches: 23467                                                                                                                                     |
|                           ->  Index Only Scan using image_tags_image_id_key on image_tags it3  (cost=0.43..0.60 rows=8 width=8) (actual time=0.002..0.003 rows=4 loops=31872)                 |
|                                 Index Cond: (image_id = it2.image_id)                                                                                                                         |
|                                 Heap Fetches: 103714                                                                                                                                          |
|                     ->  Hash  (cost=4432.36..4432.36 rows=225548 width=19) (actual time=59.026..59.026 rows=225730 loops=1)                                                                   |
|                           Buckets: 262144  Batches: 1  Memory Usage: 13849kB                                                                                                                  |
|                           ->  Seq Scan on tags t3  (cost=0.00..4432.36 rows=225548 width=19) (actual time=0.015..27.672 rows=225730 loops=1)                                                  |
|                                 Filter: (tag <> 'tagme'::text)                                                                                                                                |
|                                 Rows Removed by Filter: 1                                                                                                                                     |
| Planning Time: 1.854 ms                                                                                                                                                                       |
| Execution Time: 251.076 ms                                                                                                                                                                    |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

```
> explain analyse SELECT t3.tag AS tag, t3.count AS calc_count, it3.tag_id
                                  FROM image_tags AS it1 -- Starting image's tags
                                      INNER JOIN tags AS t1 ON t1.id = it1.tag_id AND t1.tag NOT LIKE 'tagme%'
                                      -- Get images with the same tags as the starting image
                                      INNER JOIN image_tags AS it2 ON it1.tag_id=it2.tag_id
                                      -- Get the tags from those other images except the same as the starting tags
                                      INNER JOIN image_tags AS it3 ON it2.image_id=it3.image_id
                                      LEFT JOIN image_tags it4 ON it4.image_id = it1.image_id AND it4.tag_id = it3.tag_id
                                      INNER JOIN tags AS t3 ON t3.id = it3.tag_id AND t3.tag NOT LIKE 'tagme%'
                                  WHERE
                                      it1.image_id=245368 GROUP BY it3.tag_id, t3.tag, t3.count
                                  ORDER BY calc_count DESC
                                  LIMIT 10;

+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| QUERY PLAN                                                                                                                                                                                    |
|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Limit  (cost=1751.33..1751.35 rows=10 width=19) (actual time=506.187..506.189 rows=10 loops=1)                                                                                                |
|   ->  Sort  (cost=1751.33..1757.97 rows=2657 width=19) (actual time=506.186..506.186 rows=10 loops=1)                                                                                         |
|         Sort Key: t3.count DESC                                                                                                                                                               |
|         Sort Method: top-N heapsort  Memory: 26kB                                                                                                                                             |
|         ->  HashAggregate  (cost=1667.34..1693.91 rows=2657 width=19) (actual time=505.253..505.743 rows=3223 loops=1)                                                                        |
|               Group Key: t3.count, it3.tag_id, t3.tag                                                                                                                                         |
|               ->  Nested Loop  (cost=2.15..1647.41 rows=2657 width=19) (actual time=0.045..455.517 rows=136565 loops=1)                                                                       |
|                     ->  Nested Loop  (cost=1.73..475.26 rows=2657 width=4) (actual time=0.038..128.547 rows=136996 loops=1)                                                                   |
|                           ->  Nested Loop  (cost=1.29..233.20 rows=354 width=8) (actual time=0.032..17.874 rows=31872 loops=1)                                                                |
|                                 ->  Nested Loop  (cost=0.85..87.03 rows=8 width=12) (actual time=0.020..0.029 rows=3 loops=1)                                                                 |
|                                       ->  Index Only Scan using image_tags_image_id_key on image_tags it1  (cost=0.43..19.51 rows=8 width=8) (actual time=0.009..0.011 rows=3 loops=1)        |
|                                             Index Cond: (image_id = 245368)                                                                                                                   |
|                                             Heap Fetches: 3                                                                                                                                   |
|                                       ->  Index Scan using tags_pkey on tags t1  (cost=0.42..8.44 rows=1 width=4) (actual time=0.004..0.004 rows=1 loops=3)                                   |
|                                             Index Cond: (id = it1.tag_id)                                                                                                                     |
|                                             Filter: (tag !~~ 'tagme%'::text)                                                                                                                  |
|                                 ->  Index Only Scan using image_tags_tag_id_image_id_idx on image_tags it2  (cost=0.43..12.17 rows=610 width=8) (actual time=0.010..4.832 rows=10624 loops=3) |
|                                       Index Cond: (tag_id = t1.id)                                                                                                                            |
|                                       Heap Fetches: 23467                                                                                                                                     |
|                           ->  Index Only Scan using image_tags_image_id_key on image_tags it3  (cost=0.43..0.60 rows=8 width=8) (actual time=0.002..0.003 rows=4 loops=31872)                 |
|                                 Index Cond: (image_id = it2.image_id)                                                                                                                         |
|                                 Heap Fetches: 103714                                                                                                                                          |
|                     ->  Index Scan using tags_pkey on tags t3  (cost=0.42..0.44 rows=1 width=19) (actual time=0.002..0.002 rows=1 loops=136996)                                               |
|                           Index Cond: (id = it3.tag_id)                                                                                                                                       |
|                           Filter: (tag !~~ 'tagme%'::text)                                                                                                                                    |
|                           Rows Removed by Filter: 0                                                                                                                                           |
| Planning Time: 2.247 ms                                                                                                                                                                       |
| Execution Time: 506.258 ms                                                                                                                                                                    |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
EXPLAIN
Time: 0.527s
```


```
> \d image_tags
+----------+---------+-------------+
| Column   | Type    | Modifiers   |
|----------+---------+-------------|
| image_id | integer |  not null   |
| tag_id   | integer |  not null   |
+----------+---------+-------------+
Indexes:
    "image_tags_image_id_key" UNIQUE CONSTRAINT, btree (image_id, tag_id)
    "image_tags_tag_id_image_id_idx" UNIQUE, btree (tag_id, image_id)
    "image_tags_image_id_idx" btree (image_id) CLUSTER
    "image_tags_tag_id_idx" btree (tag_id)
Foreign-key constraints:
    "image_tags_image_id" FOREIGN KEY (image_id) REFERENCES images(id) ON DELETE CASCADE
    "image_tags_tag_id" FOREIGN KEY (tag_id) REFERENCES tags(id) ON DELETE RESTRICT
```